### ISSUE TITLE
consumer.unregister hangs forever when called during a channel reconnection

### ISSUE THREAD

**Issue by thomasthebaker** [thomsonreuters/Elektron-SDK#76]
We've found an edge case condition where the call to unregister a handle in the OmmConsumerImpl will block indefinitely if called when the channel is closed. 

The logs below show our application resubscribing to items at the time of a ChannelDownReconnecting event. The ResubscriptionService thread never returns from the call to OmmConsumer.unregister. The logs messages from the Dispatcher show that the Channel went down just before the unregister was called and then came back again afterwards. Note, the channel down event was expected because we restarted the ADS server at that time. Also after the ChannelUp event the active subscriptions were fine and we were receiving data. The issue was that the ResubscriptionService thread that called the unregister method never logged any further messages but was still alive.

```
2018-08-02 10:17:55,733 INFO  [ResubscriptionService] ElektronSubscriberService - Resubscribing to FOO
2018-08-02 10:17:55,733 INFO  [ResubscriptionService] SubscriptionManager - Unregistering all clients for FOO
2018-08-02 10:17:55,733 INFO  [ResubscriptionService] SubscriptionManager - Registering client for FOO
2018-08-02 10:21:05.420 WARN  [ElektronDispatcher-Consumer] [EMA] access.OmmConsumerImpl - loggerMsg| ClientName: ChannelCallbackClient| Severity: Warning| Text:    Received ChannelDownReconnecting event on channel Channel2| RsslReactor @68be6f7c| RsslChannel @4d293eef| Error Id 0| Internal sysError 0| Error Location null| Error text SocketChannel.read returned -1 (end-of-stream)| loggerMsgEnd|
2018-08-02 10:21:05.420 WARN  [ElektronDispatcher-Consumer] [EMA] access.OmmConsumerImpl - loggerMsg| ClientName: LoginCallbackClient| Severity: Warning| Text:    RDMLogin stream state was changed to suspect with status message| username <not set>| usernameType <not set>| State: Open/Suspect/None - text: ""| loggerMsgEnd|
2018-08-02 10:21:06,079 INFO  [ResubscriptionService] ElektronSubscriberService - Resubscribing to BAR
2018-08-02 10:21:06,079 INFO  [ResubscriptionService] SubscriptionManager - Unregistering all clients for BAR
2018-08-02 10:21:06.505 INFO  [ElektronDispatcher-Consumer] [EMA] access.OmmConsumerImpl - loggerMsg| ClientName: ChannelCallbackClient| Severity: Info| Text:    Received ChannelUp event on channel Channel1| Instance Name AHLConsumer_1| Component Version ads2.6.9.L1.linux.tis.rrg 64-bit| loggerMsgEnd|
```
Note: I have replaced the item names with FOO and BAR. You can see that the FOO resubscription was successful whilst the BAR resubscription blocked the thread.

**Comment by bberner** [comment]
Thank you for your feedback. A JIRA issue has been created and the issue is being investigated. We'll let you know if we need further information.

**Comment by thomasthebaker** [comment]
I've added a test to recreate within our application, I thought it was a thread blocking from our live application logs but I have an exception stack trace that shows it's a ConcurrentModificationException in WlItemHandler:

```
2018-08-09 17:05:08,723 ERROR [pool-18-thread-1] pubsub.elektron.ElektronPubSubResiliencyIntegrationTest - Exception whilst resubscribing!
java.util.ConcurrentModificationException: null
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1442) ~[?:1.8.0_162]
	at java.util.HashMap$KeyIterator.next(HashMap.java:1466) ~[?:1.8.0_162]
	at com.thomsonreuters.upa.valueadd.reactor.WlItemHandler.removeUserRequestFromClosedStream(WlItemHandler.java:1538) ~[upaValueAdd-3.2.0.1.jar:etaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.upa.valueadd.reactor.WlItemHandler.submitMsg(WlItemHandler.java:1343) ~[upaValueAdd-3.2.0.1.jar:etaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.upa.valueadd.reactor.Watchlist.submitMsg(Watchlist.java:186) ~[upaValueAdd-3.2.0.1.jar:etaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.upa.valueadd.reactor.ReactorChannel.submit(ReactorChannel.java:558) ~[upaValueAdd-3.2.0.1.jar:etaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.ema.access.SingleItem.rsslSubmit(ItemCallbackClient.java:3000) ~[ema-3.2.0.1.jar:emaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.ema.access.SingleItem.close(ItemCallbackClient.java:2854) ~[ema-3.2.0.1.jar:emaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.ema.access.ItemCallbackClient.unregister(ItemCallbackClient.java:2256) ~[ema-3.2.0.1.jar:emaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.ema.access.OmmBaseImpl.unregister(OmmBaseImpl.java:423) ~[ema-3.2.0.1.jar:emaj3.2.0.L1.all.rrg]
	at com.thomsonreuters.ema.access.OmmConsumerImpl.unregister(OmmConsumerImpl.java:150) ~[ema-3.2.0.1.jar:emaj3.2.0.L1.all.rrg]
	at com.mi.ahl.pubsub.elektron.SubscriptionManager.unregister(SubscriptionManager.java:180) ~[classes/:?]
	at com.mi.ahl.pubsub.elektron.ElektronSubscriberService.resubscribe(ElektronSubscriberService.java:170) ~[classes/:?]
	at com.mi.ahl.pubsub.elektron.ElektronPubSubResiliencyIntegrationTest.lambda$testResubscribeWhilstChannelDown$6(ElektronPubSubResiliencyIntegrationTest.java:265) ~[test-classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_162]
	at java.util.concurrent.FutureTask.runAndReset$$$capture(FutureTask.java:308) [?:1.8.0_162]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java) [?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_162]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_162]
```
https://github.com/thomsonreuters/Elektron-SDK/blob/master/Java/Eta/ValueAdd/src/main/java/com/thomsonreuters/upa/valueadd/reactor/WlItemHandler.java#L1540

The code in the **_removeUserRequestFromClosedStream** method is not using the iterator to remove entries from the  **_pendingRequestByIdTable**. There are quite a few cases in that class that look susceptible to ConcurrentModificationException where items are removed from collections whilst iterating without using the iterator. It is probably worth reviewing all remove() usages.

**Comment by tmurthy** [comment]
This issue has been fixed in ESDK Java 1.2.2.L1. Please let us know if there are additional findings. Closing this issue
